src/lib/encoding.ts

```diff
@@ -0,0 +1,253 @@
+  let bytes = bytesFromFields(fields);
+  return new TextDecoder().decode(bytes);
+}
+
+const STOP = 0x01;
+
+function bytesToFields(bytes: Uint8Array) {
+  // we encode 248 bits (31 bytes) at a time into one field element
+  let fields = [];
+  let currentBigInt = 0n;
+  let bitPosition = 0n;
+  for (let byte of bytes) {
+    currentBigInt += BigInt(byte) << bitPosition;
+    bitPosition += 8n;
+    if (bitPosition === 248n) {
+      fields.push(Field(currentBigInt.toString()));
+      currentBigInt = 0n;
+      bitPosition = 0n;
+    }
+  }
+  // encode the final chunk, with an added STOP byte to make the mapping invertible
+  currentBigInt += BigInt(STOP) << bitPosition;
+  fields.push(Field(currentBigInt.toString()));
+  return fields;
+}
+
+function bytesFromFields(fields: Field[]) {
+  // find STOP byte in last chunk to determine length of byte array
+  let lastChunk = fields.pop();
+  if (lastChunk === undefined) return new Uint8Array();
+  let lastChunkBytes = bytesOfConstantField(lastChunk);
+  let i = lastChunkBytes.lastIndexOf(STOP, 30);
+  if (i === -1) throw Error('Error (bytesFromFields): Invalid encoding.');
+  let bytes = new Uint8Array(fields.length * 31 + i);
+  bytes.set(lastChunkBytes.subarray(0, i), fields.length * 31);
+  // convert the remaining fields
+  i = 0;
+  for (let field of fields) {
+    bytes.set(bytesOfConstantField(field).subarray(0, 31), i);
+    i += 31;
+  }
+  fields.push(lastChunk);
+  return bytes;
+}
+
+// bijective fields <--> bytes mapping
+// this is suitable for converting *arbitrary* fields AND bytes back and forth
+// the interpretation of the fields/bytes array is as digits of a single big integer
+// which implies the small caveat that trailing zeroes in the field/bytes array get ignored
+// another caveat: the algorithm is O(n^(1 + t)) with t > 0; ~1MB of field elements take about 1-2s to convert
+
+// this needs the exact field size
+let p = 0x40000000000000000000000000000000224698fc094cf91b992d30ed00000001n;
+let q = 0x40000000000000000000000000000000224698fc0994a8dd8c46eb2100000001n;
+let bytesPerBigInt = 32;
+let bytesBase = 256n ** BigInt(bytesPerBigInt);
+
+const Bijective = {
+  Fp: {
+    toBytes: (fields: Field[]) => toBytesBijective(fields, p),
+    fromBytes: (bytes: Uint8Array) => toFieldsBijective(bytes, p),
+
+    toString(fields: Field[]) {
+      return new TextDecoder().decode(toBytesBijective(fields, p));
+    },
+    fromString(message: string) {
+      let bytes = new TextEncoder().encode(message);
+      return toFieldsBijective(bytes, p);
+    },
+  },
+  Fq: {
+    toBytes: (fields: Field[]) => toBytesBijective(fields, q),
+    fromBytes: (bytes: Uint8Array) => toFieldsBijective(bytes, q),
+
+    toString(fields: Field[]) {
+      return new TextDecoder().decode(toBytesBijective(fields, q));
+    },
+    fromString(message: string) {
+      let bytes = new TextEncoder().encode(message);
+      return toFieldsBijective(bytes, q);
+    },
```

Comment:
would be nice to have some sanity check tests to make sure you can go back and forth

Reply:
Yeah I still need to find a testing flow that feels nice and works well with our JS setup.. Right now I'm adding "tests" as examples that I manually run to see if it works

Reply:
this is the file where it's "tested" - src/examples/encoding-bijective.ts